### PR TITLE
Experimental merge queue

### PR DESCRIPTION
**Pull request by casperisfine**
This implements a simple merge queue for stacks.
  - Pull Requests registered
  - If the stack allows merges (is not locked, is not backlogged, is not failing), the pull request will be merged.
  - If the pull request is not mergeable (merge conflict, failing commit status, etc) it will be rejected.
  - Pull requests can be removed from the queue.
  - API endpoints to list / add / remove pull requests from the queue.

Future features:
  - UI to add a PR to the queue
  - UI to list PRs in the queue

### CODE REVIEW

**Comment by casperisfine** [comment]
cc @DazWorrall @wvanbergen @pior @etiennebarrie 

The PR is a bit massive and you don't have all the context on Shipit, so I don't expect a deep review. But it could be worth it to give it a quick glance.

The key parts being `MergePullRequestsJob` and the `PullRequest` model.

File: app/models/shipit/pull_request.rb [hunk 1]
```diff
@@ -0,0 +1,184 @@
+module Shipit
+  class PullRequest < ApplicationRecord
```
Comment (wvanbergen) [inline app/models/shipit/pull_request.rb]: Do we keep track of all of them? Or just the ones that were requested to be merged?

File: app/models/shipit/pull_request.rb [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    belongs_to :head, class_name: 'Shipit::Commit'
+    belongs_to :merge_requested_by, class_name: 'Shipit::User'
+
+    validates :number, presence: true, uniqueness: {scope: :stack_id}
+
+    scope :waiting, -> { where(merge_status: WAITING_STATUSES) }
+    scope :pending, -> { where(merge_status: 'pending') }
+    scope :to_be_merged, -> { pending.order(merge_requested_at: :asc) }
+
+    state_machine :merge_status, initial: :fetching do
```
Comment (wvanbergen) [inline app/models/shipit/pull_request.rb]: I wonder if it makes sense to have some docs about this state machine and the process behind merging a PR available.

File: app/models/shipit/unlimited_api_client.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module Shipit
+  class UnlimitedApiClient
```
Comment (wvanbergen) [inline app/models/shipit/unlimited_api_client.rb]: This is for development purposes?

File: test/controllers/api/commits_controller_test.rb [hunk 4]
```diff
@@ -9,7 +9,7 @@ class CommitsControllerTest < ActionController::TestCase
       end
 
       test "#index returns a list of commits" do
-        commit = @stack.commits.last
+        commit = @stack.commits.reachable.last
```
Comment (wvanbergen) [inline test/controllers/api/commits_controller_test.rb]: Is this related to supporting PRs? Should be have a test for this?

File: app/models/shipit/pull_request.rb (hunk 2 above)
Comment (casperisfine) [inline app/models/shipit/pull_request.rb]: Ultimately it will be documented in the `README.md`, but for now it's an experimental feature marked as alpha that is likely to change in the future. It won't be documented until it's dimmed stable enough.

File: app/models/shipit/unlimited_api_client.rb (hunk 3 above)
Comment (casperisfine) [inline app/models/shipit/unlimited_api_client.rb]: Yes, I was tired of crafting queries. It's optionally enabled with an environment variable and disabled by default.

File: test/controllers/api/commits_controller_test.rb (hunk 4 above)
Comment (casperisfine) [inline test/controllers/api/commits_controller_test.rb]: It's a side effect. Commits in Shipit have a `detached: boolean` attribute. When someone force push a tracked branch, some commits we recorded are not on any branch anymore, and we mark them as such.

And since this only happen on force pushes, the last commit of a stack was incidentally never detached.

But Pull Request do reuse the commit model, they point to a `head` that is a detached commit. Because of this it is now possible for the last commit to be detached which broke that assumption, and several tests.

**Comment by casperisfine** [comment]
> Is this for the simple use case (as defined by our RFC), where it simply immediately merges a PR if allowed when a request comes in, and fail otherwise?

Because the process have to be asynchronous anyway, as it's not acceptable to reach to the GitHub API inline.

**Comment by wvanbergen** [comment]
> Because the process have to be asynchronous anyway, as it's not acceptable to reach to the GitHub API inline. 

Makes sense. It also sets us up for later versions where we may not immediately reject, but wait for a suitable merge window.

**Comment by casperisfine** [comment]
> Makes sense. It also sets us up for later versions where we may not immediately reject, but wait for a suitable merge window.

Well, this version only reject PRs that are not mergeable by nature (CI failing or pending, merge conflict, etc).

But if it can't be merged because of externalities (stack locked, branch failing or backlogged), then Shipit will wait for a suitable merge window.

One thing that could change is the rejection of PRs with pending CI status. I though that I don't want people to put PRs in the queue before they are ready, but one could argue it makes sense (i.e. Ship this ASAP please).

**Comment by wvanbergen** [comment]
> But if it can't be merged because of externalities (stack locked, branch failing or backlogged), then Shipit will wait for a suitable merge window.

Should we put a timeout on this? E.g. maximum an hour.

> One thing that could change is the rejection of PRs with pending CI status. I though that I don't want people to put PRs in the queue before they are ready, but one could argue it makes sense (i.e. Ship this ASAP please).

We also have some cases where we want to get PRs (especially reverts) to be deployed ASAP. We can just work around the queue for those cases just now, but might be nice to have an override available for those cases.


Something that is a bit hard for me to figure out is what happens when the stack gets locked. Should we cancel all merge requests?

**Comment by casperisfine** [comment]
> Should we put a timeout on this? E.g. maximum an hour.

That's a good idea. I'd do that in a followup though.

> We also have some cases where we want to get PRs (especially reverts) to be deployed ASAP. We can just work around the queue for those cases just now, but might be nice to have an override available for those cases.

IMO in that scenario we should stick with manual merges. If we need a PR to go out ASAP, I don't see the point asking Shipit when it thinks it's be best to merge it.

Also Shipit is entirely built around the idea of not preventing behaviors, only honest mistakes, so personally I'm not really fond of the idea of restricting merge permissions. But I'm open to it if early experiments show that people can't be trusted.

> Something that is a bit hard for me to figure out is what happens when the stack gets locked. Should we cancel all merge requests?

Nope. The entire point of the queue is to reduce the impact of externalities on the shipping pipeline. If we were to invalidate all the request son lock, that wouldn't be very effective.

So in this current implementation, nothing will be merged while the lock is in place, but as soon as it's released, merging will resume.

**Comment by wvanbergen** [comment]
Initial rollout plan:

1. Add spy command.
2. Tell people they can use this spy command instead of merging themselves?

???

**Comment by casperisfine** [comment]
Yes, I already started the spy integration. But I think the best would be HCTW integration, and it's actually even simpler to implement.

File: app/models/shipit/stack.rb [hunk 5]
```diff
@@ -179,19 +181,23 @@ def head
     end
 
     def merge_status
-      if locked?
-        'locked'
+      return 'locked' if locked?
+      return 'failure' if branch_status == 'failure'
+      if undeployed_commits_count > (maximum_commits_per_deploy || DEFAULT_MAX_COMMITS_PER_DEPLOY) * 1.5
```
Comment (etiennebarrie) [inline app/models/shipit/stack.rb]: Do you think it would be worthwhile to override `maximum_commits_per_deploy` to have the default there?

File: test/fixtures/shipit/pull_requests.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  merge_status: pending
+  merge_requested_at: <%= 3.minute.ago.to_s(:db) %>
+  github_id: 45454454545454
+  api_url: https://api.github.com/repos/shopify/shipit-engine/pulls/64
+  state: open
+  head_id: 8
+  mergeable: null
+  additions: 23
+  deletions: 43
+
```
Comment (etiennebarrie) [inline test/fixtures/shipit/pull_requests.yml]: blank line

File: test/fixtures/shipit/statuses.yml [hunk 7]
```diff
@@ -90,3 +90,12 @@ cyclimse_success_travis:
...
 
+shipit_pending_pr_success_travis:
+  stack: shipit
+  commit_id: 8 # shipit_pending_pr_head
+  description: Super feature
+  context: ci/travis
+  created_at: <%= 9.days.ago.to_s(:db) %>
+  state: success
+  target_url: "http://www.example.com"
+
```
Comment (etiennebarrie) [inline test/fixtures/shipit/statuses.yml]: whitespace

**Comment by casperisfine** [comment]
> I think we can go further and remove the hidden? and required? methods in Status::Common. allowed_to_fail? is still used it in the view, but it will probably have to change for the queue UI (to show the proper allowed to fail statuses as ignored).

Yeah I'm not super happy with that either. In the end it's because of the disparity between "branch" commits and "PR head" commits. So these methods should probably move in some kind of decorator. But as you said, not a huge deal right now.

Thanks for the review!
